### ISSUE TITLE
Update S3440: Allow property check before assignment

### ISSUE THREAD

**Issue by ValentijnMakkenze** [SonarSource/sonar-dotnet#2313]
### Description
We get a false positive for Remove this useless conditional.

### Repro steps

In the following example we guard against assigning the same  value to a property with side affects.

According to the documentation:
Properties are excluded from this rule because they could have side effects and removing the check could lead to undesired side effects.

```csharp
namespace WindowsFormsApp3
{
  public class ClassWithProperty
  {
    private string myProperty;

    private void DoIncridibleExpensiveOperationHere(string argument)
    {
      myProperty = argument;
    }

    public string MyProperty
    {
      get
      {
        return myProperty;
      }

      set
      {
        DoIncridibleExpensiveOperationHere(value);
      }
    }
  }

  public class S3440Test
  {
    private readonly ClassWithProperty test = new ClassWithProperty();

    public void ReproduceWarning(string argument)
    {
      if (test.MyProperty != argument) // Remove this useless conditional.
      {
        test.MyProperty = argument;
      }
    }
  }
}
```

### Expected behavior
No warning

### Actual behavior
Warning: Remove this useless conditional.

### Known workarounds
Mark item as false positive

### Related information
SonarC# Version 7.11.0.8083
Visual Studio Version 2017
If running through the Scanner for MSBuild, its version 4.5.0.1761

Reproduced in VS2019 with SonarAnalyzer.CSharp 7.11.0.8083

**Comment by kvpt** [comment]
I also noticed that after updating the dotnet analyser.
The code of the rule that detect if it's a property affectation is wrong.
I made a PR to fix that.

**Comment by Evangelink** [comment]
Hi @ValentijnMakkenze, @kvpt,

Thank you for the feedback! 

We were actually checking properties but the sentence was unclear to about what we were allowing. We wanted to allow field check before the assignment inside the setter because there might be some expansive operation done in this case. While you understand the sentence as checking the property value outside the setter.

I do think that both options are equally valid so I will update the rule definition and its code.

Cheers,
Amaury
